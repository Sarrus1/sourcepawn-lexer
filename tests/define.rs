mod utils;

use insta::assert_json_snapshot;
use sourcepawn_lexer::*;
use utils::collect_tokens;

#[test]
fn define_simple() {
    let input = r#"#define FOO 1
     "#;

    let mut lexer = SourcepawnLexer::new(input);
    assert_json_snapshot!(collect_tokens(&mut lexer));
}

#[test]
fn define_no_value() {
    let input = r#"#define FOO
"#;

    let mut lexer = SourcepawnLexer::new(input);
    assert_json_snapshot!(collect_tokens(&mut lexer));
}

#[test]
fn define_no_line_break() {
    let input = "#define FOO 1";

    let mut lexer = SourcepawnLexer::new(input);
    assert_json_snapshot!(collect_tokens(&mut lexer));
}

#[test]
fn define_trailing_line_comment() {
    let input = r#"#define FOO 1 //bar
"#;

    let mut lexer = SourcepawnLexer::new(input);
    assert_json_snapshot!(collect_tokens(&mut lexer));
}

#[test]
fn define_trailing_block_comment() {
    let input = r#"#define FOO 1 /* */
"#;

    let mut lexer = SourcepawnLexer::new(input);
    assert_json_snapshot!(collect_tokens(&mut lexer));
}

#[test]
fn define_with_block_comment() {
    let input = r#"#define FOO 1 /* */ + 1
"#;

    let mut lexer = SourcepawnLexer::new(input);
    assert_json_snapshot!(collect_tokens(&mut lexer));
}

#[test]
fn define_with_block_comment_and_line_continuation() {
    let input = r#"#define FOO 1 /* */ \
+ 1
"#;

    let mut lexer = SourcepawnLexer::new(input);
    assert_json_snapshot!(collect_tokens(&mut lexer));
}

#[test]
fn define_with_trailing_multiline_block_comment() {
    let input = r#"#define FOO 1 /*
*/ + 1
"#;

    let mut lexer = SourcepawnLexer::new(input);
    assert_json_snapshot!(collect_tokens(&mut lexer));
}

#[test]
fn define_with_trailing_line_continuated_multiline_block_comment() {
    let input = r#"#define FOO 1 /* \
*/ + 1
"#;

    let mut lexer = SourcepawnLexer::new(input);
    assert_json_snapshot!(collect_tokens(&mut lexer));
}

#[test]
fn define_line_continuation() {
    let input = r#"#define FOO 1 \
+ 1
"#;

    let mut lexer = SourcepawnLexer::new(input);
    assert_json_snapshot!(collect_tokens(&mut lexer));
}

#[test]
fn define_line_continuation_carriage_return() {
    let input = "#define FOO 1 \\\r\n+ 1\n";

    let mut lexer = SourcepawnLexer::new(input);
    assert_json_snapshot!(collect_tokens(&mut lexer));
}
