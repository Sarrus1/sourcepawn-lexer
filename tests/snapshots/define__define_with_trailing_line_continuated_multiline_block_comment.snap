---
source: tests/define.rs
assertion_line: 87
expression: collect_tokens(&mut lexer)
snapshot_kind: text
---
[
  {
    "kind": {
      "PreprocDir": "MDefine"
    },
    "text": "#define",
    "range_start": 0,
    "range_end": 7,
    "delta": 0,
    "in_preprocessor": true
  },
  {
    "kind": "Identifier",
    "text": "FOO",
    "range_start": 8,
    "range_end": 11,
    "delta": 1,
    "in_preprocessor": true
  },
  {
    "kind": {
      "Literal": "IntegerLiteral"
    },
    "text": "1",
    "range_start": 12,
    "range_end": 13,
    "delta": 1,
    "in_preprocessor": true
  },
  {
    "kind": {
      "Comment": "BlockComment"
    },
    "text": "/* \\\n*/",
    "range_start": 14,
    "range_end": 21,
    "delta": 1,
    "in_preprocessor": true
  },
  {
    "kind": {
      "Operator": "Plus"
    },
    "text": "+",
    "range_start": 22,
    "range_end": 23,
    "delta": 1,
    "in_preprocessor": true
  },
  {
    "kind": {
      "Literal": "IntegerLiteral"
    },
    "text": "1",
    "range_start": 24,
    "range_end": 25,
    "delta": 1,
    "in_preprocessor": true
  },
  {
    "kind": "Newline",
    "text": "\n",
    "range_start": 25,
    "range_end": 26,
    "delta": 0,
    "in_preprocessor": false
  },
  {
    "kind": "Eof",
    "text": "\u0000",
    "range_start": 26,
    "range_end": 26,
    "delta": 0,
    "in_preprocessor": false
  }
]
